prog/vnet/load_balancer_nexus.rb

```diff
@@ -43,11 +46,72 @@ def before_run
       decr_rewrite_dns_records
     end
 
+    when_refresh_cert_set? do
+      hop_refresh_cert
+    end
+
     hop_create_new_health_probe if strand.children_dataset.count < load_balancer.vms_dataset.count
 
     nap 5
   end
 
+  label def refresh_cert
+    private_key = OpenSSL::PKey::RSA.new(4096)
+    client = Acme::Client.new(private_key: private_key, directory: Config.lets_encrypt_directory)
+    account = client.new_account(contact: "mailto:#{Config.lets_encrypt_email}", terms_of_service_agreed: true)
+    order = client.new_order(identifiers: [load_balancer.hostname])
+    authorization = order.authorizations.first
+    load_balancer.update(kid: account.kid, private_key: private_key, order_url: order.url)
+    dns_challenge = authorization.dns
+    Prog::Vnet::LoadBalancerNexus.dns_zone.insert_record(record_name: "#{dns_challenge.record_name}.#{load_balancer.hostname}", type: dns_challenge.record_type, ttl: 600, data: dns_challenge.record_content)
+
+    hop_wait_dns_update
+  end
+
+  label def wait_dns_update
+    dns_challenge.reload
+
+    dns_record = DnsRecord[dns_zone_id: Prog::Vnet::LoadBalancerNexus.dns_zone.id, name: "#{dns_challenge.record_name}.#{load_balancer.hostname}.", tombstoned: false]
+    if DB[:seen_dns_records_by_dns_servers].where(dns_record_id: dns_record.id).empty?
+      nap 10
+    end
+
+    dns_challenge.request_validation
+
+    hop_wait_dns_validation
+  end
+
+  label def wait_dns_validation
+    dns_challenge.reload
+    if dns_challenge.status == "pending"
+      nap 10
+    elsif dns_challenge.status == "valid"
+      csr_key = OpenSSL::PKey::RSA.new(4096)
+      csr = Acme::Client::CertificateRequest.new(private_key: csr_key, common_name: load_balancer.hostname)
+      acme_order.finalize(csr: csr)
+      load_balancer.update(csr_key: csr_key.to_pem)
+
+      hop_wait_cert_finalization
+    else
+      fail "DNS validation failed"
+    end
+  end
+
+  label def wait_cert_finalization
+    acme_order.reload
+    if acme_order.status == "processing"
+      nap 10
+    elsif acme_order.status == "valid"
+      load_balancer.update(cert: acme_order.certificate, cert_created_at: Time.now)
+      puts "DNS RECORD: #{dns_challenge.record_name}.#{load_balancer.hostname}"
+      Prog::Vnet::LoadBalancerNexus.dns_zone.delete_record(record_name: "#{dns_challenge.record_name}.#{load_balancer.hostname}")
+      decr_refresh_cert
+      hop_wait
+    else
+      fail "Certificate finalization failed"
+    end
+  end
+
   label def create_new_health_probe
     vms = load_balancer.vms
     vms_getting_probed = strand.children_dataset.where(prog: "Vnet::LoadBalancerHealthProbes").map { |st| st.stack[0]["vm_id"] }
```

Comment:
just noticing this now, I know it's not part of this patch, but it's an oof given it'll break under `push`. Also oofy is crawling around in the child's data structure, it seems like any monotonic logic here is hard to analyze and keep intact.

Reply:
So, instead of creating health checks here, I can create them at the time of `add_vm`. I wanted to be able to backfill health checks just in case an operator (me), messes up.

Reply:
I have implemented this in #1828